server/src/main/java/io/crate/metadata/sys/TableHealthService.java

```diff
@@ -165,7 +167,23 @@ private TableHealth tableHealthFromEntry(Map.Entry<TablePartitionIdent, ShardsIn
         } catch (RelationUnknown e) {
             return null;
         }
-        return calculateHealth(ident, shardsInfo, tableInfo.numberOfShards());
+        int shardsCount = tableInfo.numberOfShards();
+        if (ident.partitionIdent != null) {
+            var partitionIndexName = IndexParts.toIndexName(
+                new RelationName(ident.tableSchema, ident.tableName),
+                ident.partitionIdent
+            );
+            var indexMetaData = clusterService.state().getMetadata().index(partitionIndexName);
+            assert indexMetaData != null : "index meta data for partition " + partitionIndexName + " not found";
+            var partitionInfo = PartitionInfos.createPartitionInfo(
+                partitionIndexName,
+                indexMetaData
+            );
+            assert partitionInfo != null : "cannot resolve partition info for " + partitionIndexName;
+            shardsCount = partitionInfo.numberOfShards();
```

Comment:
```suggestion
            shardsCount = indexMetadata.getNumberOfShards();
```

No need to go through the `PartitionInfo`  to access the shard count.